File: src/mask.jl
using Images

# type Mask struct {
#   Start    int64 `json:"start,omitempty"`
#   End      int64 `json:"end,omitempty"`
#   BBox     BBox  `json:"bbox,omitempty"`
#   NullMask bool  `json:"null_mask,omitempty"`
# }

function mask_img(A::Image, masks::Array)
    fill = zero(eltype(A.data))
    for m in masks
        if haskey(m, "null_mask")
            continue
        end
        if haskey(m, "start")
            println("masking [:, $(m["start"]):$(m["end"])]")
            A.data[:, m["start"]:m["end"]] = fill
        else
            bbox = m["bbox"]
            println("masking [$(bbox["a"]["x"]):$(bbox["b"]["x"]), $(bbox["a"]["y"]):$(bbox["b"]["y"])]")
            A.data[bbox["a"]["x"]:bbox["b"]["x"], bbox["a"]["y"]:bbox["b"]["y"]] = fill
        end
    end
    A
end 

function mask(A::BitMatrix, masks::Array)
    for m in masks
        if haskey(m, "null_mask")
            continue
        end
        if haskey(m, "start")
            A[:, m["start"]:m["end"]] = false
        else
            bbox = m["bbox"]
            A[bbox["a"]["x"]:bbox["b"]["x"], bbox["a"]["y"]:bbox["b"]["y"]] = false
        end
    end
    A
end

File: src/lib/vz_project/mask.go
package project

import (
	"fmt"
	"image"
	"image/color"
	"image/draw"
	"log"
	"math"
	"strconv"
	"strings"
)

type Mask struct {
	Start    int64 `json:"start,omitempty"`
	End      int64 `json:"end,omitempty"`
	BBox     *BBox `json:"bbox,omitempty"`
	NullMask bool  `json:"null_mask,omitempty"`
}

func (m Mask) String() string {
	if m.NullMask {
		return "-"
	}
	if m.Start != 0 && m.End != 0 {
		return fmt.Sprintf("%d:%d", m.Start, m.End)
	}
	return m.BBox.String()
}

func ParseMask(s string) (m Mask, ok bool) {
	s = strings.TrimSpace(s)
	switch {
	case s == "-":
		m.NullMask = true
		ok = true
	case strings.Count(s, ":") == 1:
		c := strings.SplitN(s, ":", 2)
		x, _ := strconv.Atoi(c[0])
		y, _ := strconv.Atoi(c[1])
		m.Start = int64(math.Min(float64(x), float64(y)))
		m.End = int64(math.Max(float64(x), float64(y)))
		ok = true
	case strings.Count(s, "x") == 2 && strings.Count(s, " ") == 1:
		m.BBox = ParseBBox(s)
		ok = !m.BBox.IsZero()
	}
	return
}

var black = image.NewUniform(color.Gray{})

type Masks []Mask

func (m Masks) Apply(i image.Image) {
	var ii draw.Image
	var ok bool
	if ii, ok = i.(draw.Image); !ok {
		log.Printf("%T does not implement draw.Image")
		return
	}
	for _, mm := range m {
		var r image.Rectangle
		if mm.BBox != nil {
			r = mm.BBox.Rect()
		} else {
			r.Min.Y = int(mm.Start)
			r.Max.Y = int(mm.End)
			r.Max.X = i.Bounds().Dx()
		}
		// log.Printf("masking %v", r)

		// adjust by .Min{X,Y} which might not be zero
		r = r.Add(ii.Bounds().Min)
		draw.Draw(ii, r, black, image.ZP, draw.Src)
	}
}
